src/user-safe/UserSafe.sol

```diff
@@ -0,0 +1,678 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {SafeERC20, IERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC20Permit} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Permit.sol";
+import {ICashDataProvider} from "../interfaces/ICashDataProvider.sol";
+import {SignatureUtils} from "../libraries/SignatureUtils.sol";
+import {ISwapper} from "../interfaces/ISwapper.sol";
+import {IPriceProvider} from "../interfaces/IPriceProvider.sol";
+import {OwnableUpgradeable} from "openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import {Initializable} from "openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol";
+
+/**
+ * @title UserSafe
+ * @author ether.fi [shivam@ether.fi]
+ * @notice User safe account for interactions with the EtherFi Cash contracts
+ */
+contract UserSafe is Initializable, OwnableUpgradeable {
+    using SafeERC20 for IERC20;
+    using SignatureUtils for bytes32;
+
+    enum SpendingLimitTypes {
+        Daily,
+        Weekly,
+        Monthly,
+        Yearly
+    }
+
+    struct WithdrawalRequest {
+        address[] tokens;
+        address recipient;
+        uint96 finalizeTime;
+    }
+
+    struct WithdrawalData {
+        address[] tokens;
+        uint256[] amounts;
+        address recipient;
+        uint96 finalizeTime;
+    }
+
+    struct SpendingLimitData {
+        SpendingLimitTypes spendingLimitType;
+        uint64 renewalTimestamp;
+        uint256 spendingLimit; // in USD with 6 decimals
+        uint256 usedUpAmount; // in USD with 6 decimals
+    }
+
+    bytes32 public constant REQUEST_WITHDRAWAL_METHOD =
+        keccak256("requestWithdrawal");
+    bytes32 public constant APPROVE_METHOD = keccak256("approve");
+    bytes32 public constant SET_SPENDING_LIMIT_METHOD =
+        keccak256("setSpendingLimit");
+    bytes32 public constant SET_INCOMING_SPENDING_LIMIT_METHOD =
+        keccak256("setIncomingSpendingLimit");
+
+    // Address of the USDC token
+    address public immutable usdc;
+    // Address of the weETH token
+    address public immutable weETH;
+    // Address of the Cash Data Provider
+    ICashDataProvider public immutable cashDataProvider;
+    // Address of the price provider
+    IPriceProvider public immutable priceProvider;
+    // Address of the swapper
+    ISwapper public immutable _swapper;
+    // Withdrawal requests pending with the contract
+    WithdrawalRequest private _pendingWithdrawalRequest;
+    // Funds blocked for withdrawal
+    mapping(address token => uint256 amount) private blockedFundsForWithdrawal;
+    // Nonce for permit operations
+    uint256 private _nonce;
+    // Current spending limit
+    SpendingLimitData private _spendingLimit;
+    // This spending limit gets activated once the _spendingLimit time is over
+    SpendingLimitData private _incomingSpendingLimit;
```

Comment:
can u explain why we need to maintain these two separately? any race condition we are worried about?

Reply:
These are just so that the user can set an incoming spending limit in advance after the current limit gets over. They can just modify the current limit directly as well but I thought it would be better if they could set the new spending limit in advance before the current one gets over.

Reply:
I see. but I am not sure if it is good for UX. I would keep one `_spendingLimit` instance and let user update the current one in-place. 

Better to ask hear opinion from Bryan, Koray on it

Reply:
Asked them, the final solution we are going with is 
1. no start timestamp
2. no incoming spending limits
3. renew spending limit after each interval automatically
4. a new function to update the spending limit within the same interval

Reply:
great!